File: src/main/java/com/mobilesolutionworks/android/zero/Struts.java
package com.mobilesolutionworks.android.zero;

import android.content.Context;
import android.support.annotation.NonNull;

/**
 * Created by yunarta on 9/5/17.
 */

public interface Struts {

    <T extends EndPoint.Contract> T endPoint(@NonNull Class<T> name);

    @NonNull
    Context context();

    /**
     * Created by yunarta on 9/5/17.
     */

    interface BackEnd extends Struts {

        @NonNull
        Context context();

        @NonNull
        <L extends EntityLocator.Contract> L locator(@NonNull Class<L> name);
    }

    /**
     * Created by yunarta on 9/5/17.
     */

    interface Install {

        <T extends EndPoint.Contract, M extends EndPoint<T>> void addEndPoint(@NonNull Class<T> name, @NonNull M module);

        <L extends EntityLocator.Contract, M extends EntityLocator<L>> void addLocator(@NonNull Class<L> name, @NonNull M module);
    }

    /**
     * Created by yunarta on 9/5/17.
     */

    interface Setup {

        void installPlugin(@NonNull Plugin plugin);
    }
}

File: src/main/kotlin/com/mobilesolutionworks/android/struts/Struts.kt
package com.mobilesolutionworks.android.struts

import kotlin.reflect.KClass

/**
 * Created by yunarta on 13/5/17.
 */

public interface Struts {

    fun <T : EndPoint.Contract> endPoint(name: Class<T>): T?

    operator fun <T : EndPoint.Contract> get(name: KClass<T>): T? = endPoint(name.java)

    val context: android.content.Context

    interface BackEnd<out S : Scheduler> : Struts {

        val scheduler: S

        fun <L : EntityLocator.Contract> locator(name: Class<L>): L?
    }

    interface Install {

        fun <T : EndPoint.Contract, M : EndPoint<T>> addEndPoint(name: Class<T>, endPoint: M)

        fun <L : EntityLocator.Contract, M : EntityLocator<L>> addLocator(name: Class<L>, locator: M)
    }

    interface Mutable {

        val scheduler: Scheduler.Install
    }
}
